internal/agg/aggregate.go

```diff
@@ -0,0 +1,240 @@
+	case int64:
+		target.val = target.val.(int64) + new.val.(int64)
+	case float32:
+		target.val = target.val.(float32) + new.val.(float32)
+	case float64:
+		target.val = target.val.(float64) + new.val.(float64)
+	default:
+		return fmt.Errorf("unsupported type: %T", target.val)
+	}
+	return nil
+}
+
+func AggregatesToPB(aggregates []AggregateBase) []*planpb.Aggregate {
+	ret := make([]*planpb.Aggregate, len(aggregates))
+	for idx, agg := range aggregates {
+		ret[idx] = agg.ToPB()
+	}
+	return ret
+}
+
+type FieldValue struct {
+	val interface{}
+}
+
+func NewFieldValue(v interface{}) *FieldValue {
+	return &FieldValue{val: v}
+}
+
+type Row struct {
+	fieldValues []*FieldValue
+}
+
+func (r *Row) ColumnCount() int {
+	return len(r.fieldValues)
+}
+
+func (r *Row) ValAt(col int) interface{} {
+	return r.fieldValues[col].val
+}
+
+func (r *Row) Equal(other *Row, keyCount int) bool {
+	// Ensure both rows have enough fields for key comparison
+	if len(r.fieldValues) < keyCount || len(other.fieldValues) < keyCount || len(r.fieldValues) != len(other.fieldValues) {
+		return false
+	}
+	// Compare each field value for equality
+	for i := 0; i < keyCount; i++ {
+		if r.fieldValues[i].val != other.fieldValues[i].val {
+			return false
+		}
+	}
+	return true
+}
+
+func (r *Row) UpdateFieldValue(newRow *Row, col int, agg AggregateBase) {
+	agg.Update(r.fieldValues[col], newRow.fieldValues[col])
+}
+
+func (r *Row) ToString() string {
+	var builder strings.Builder
+	builder.WriteString("agg-row:")
+	for _, fv := range r.fieldValues {
+		builder.WriteString(fmt.Sprintf("%v,", fv.val))
+	}
+	return builder.String()
+}
+
+func NewRow(fieldValues []*FieldValue) *Row {
+	return &Row{fieldValues: fieldValues}
+}
+
+type Bucket struct {
+	rows []*Row
+}
+
+func (bucket *Bucket) AddRow(row *Row) {
+	bucket.rows = append(bucket.rows, row)
+}
+
+func (bucket *Bucket) RowAt(idx int) *Row {
+	return bucket.rows[idx]
+}
+
+func (bucket *Bucket) RowCount() int {
+	return len(bucket.rows)
+}
+
+func (bucket *Bucket) Accumulate(row *Row, idx int, keyCount int, aggs []AggregateBase) error {
+	if idx >= len(bucket.rows) || idx < 0 {
+		return fmt.Errorf("wrong idx:%d for bucket", idx)
+	}
+	targetRow := bucket.rows[idx]
+	if targetRow == nil {
+		return fmt.Errorf("nil row at the target idx:%d, cannot accumulate the row", idx)
+	}
+	if row.ColumnCount() != targetRow.ColumnCount() {
+		return fmt.Errorf("column count:%d in the row must be equal to the target row:%d", row.ColumnCount(), bucket.rows[idx].ColumnCount())
+	}
+	if row.ColumnCount() != keyCount+len(aggs) {
+		return fmt.Errorf("column count:%d in the row must be sum of keyCount:%d and the number of aggs:%d", row.ColumnCount(), keyCount, len(aggs))
+	}
+	for col := keyCount; col < row.ColumnCount(); col++ {
+		targetRow.UpdateFieldValue(row, col, aggs[col-keyCount])
+	}
+	return nil
+}
+
+const NONE int = -1
+
+func (bucket *Bucket) Find(row *Row, keyCount int) int {
+	for idx, existingRow := range bucket.rows {
```

Comment:
1. why do we setup keyCount when when create bucket? so we don't need to pass keyCount.
2. should be keep map[keyHash]idx? when groupkey is a lot this could be super expensive